pkg/controller/master/addon/addon.go

```
@@ -0,0 +1,301 @@
+
+	var redeployChart bool
+	if err != nil {
+		// HelmChart is missing. Reset Addon to allow normal reconcile
+		if apierrors.IsNotFound(err) {
+			redeployChart = true
+		} else {
+			return a, fmt.Errorf("error fetching helm chart during monitor changes: %v", err)
+		}
+	}
+
+	// if chart is being deleted then but addon is not then redeploy chart
+	// this is possible since helm-controller adds a finalizer to uninstall the chart when HelmChart CRD is being deleted
+
+	if hc.DeletionTimestamp != nil {
+		redeployChart = true
+	}
+
+	if redeployChart {
+		a.Status.Status = ""
+		return h.addon.UpdateStatus(a)
+	}
+
+	vals, err := defaultValues(a)
+	if err != nil {
+		return a, fmt.Errorf("error generating default values during monitor changes: %v", err)
+	}
+	if hc.Spec.ValuesContent != vals || hc.Spec.Version != a.Spec.Version || hc.Spec.Chart != a.Spec.Chart || hc.Spec.Repo != a.Spec.Repo {
+		hc.Spec.Chart = a.Spec.Chart
+		hc.Spec.Repo = a.Spec.Repo
+		hc.Spec.Version = a.Spec.Version
+		hc.Spec.ValuesContent = vals
+		_, err = h.helm.Update(hc)
+		if err != nil {
+			return a, fmt.Errorf("error updating helm chart during monitor changes: %v", err)
+		}
+	}
+
+	// reset status to AddonEnable to trigger reconcile for job
+	a.Status.Status = harvesterv1.AddonEnabled
+	return h.addon.UpdateStatus(a)
+}
+
+// MonitorAddon will track the deployment of Addon
+func (h *Handler) MonitorAddon(key string, a *harvesterv1.Addon) (*harvesterv1.Addon, error) {
+	if a == nil || !a.DeletionTimestamp.IsZero() {
+		return nil, nil
+	}
+
+	if a.Status.Status != harvesterv1.AddonEnabled {
+		return a, nil
+	}
+
+	hc, err := h.helm.Get(a.Namespace, a.Name, metav1.GetOptions{})
+	if err != nil {
+		return a, fmt.Errorf("error querying helmchart in checkHelmChartExists: %v", err)
+	}
+
+	if hc.Status.JobName == "" {
+		return a, fmt.Errorf("waiting for jobname to be populated in HelmChart status")
+	}
+
+	job, err := h.job.Get(a.Namespace, hc.Status.JobName, metav1.GetOptions{})
```

ditto and don't swallow error